shell/server/core.js

```diff
@@ -0,0 +1,241 @@
+    // For ongoing notifications, it will begin the process of cancelling and dropping them from
+    // the app.
+    var notification = Notifications.findOne({_id: notificationId});
+    if (!notification) {
+      throw new Meteor.Error(404, "Notification id not found.");
+    } else if (notification.userId !== Meteor.userId()) {
+      throw new Meteor.Error(403, "Notification does not belong to current user.");
+    } else {
+      dismissNotification(notificationId);
+    }
+  },
+  readAllNotifications: function () {
+    // Marks all notifications as read for the current user.
+    if (!Meteor.userId()) {
+      throw new Meteor.Error(403, "User not logged in.");
+    }
+    Notifications.update({userId: Meteor.userId()}, {$set: {isUnread: false}}, {multi: true});
+  }
+});
+
+NotificationHandle.prototype.close = function () {
+  var self = this;
+  return inMeteor(function () {
+    if (!self.saved) {
+      dismissNotification(self.notificationId);
+    }
+  });
+};
+
+NotificationHandle.prototype.save = function () {
+  var self = this;
+  return inMeteor(function () {
+    var sturdyRef = new Buffer(Random.id(20));
+    var hashedSturdyRef = hashSturdyRef(sturdyRef);
+    ApiTokens.insert({
+      _id: hashedSturdyRef,
+      frontendRef: {
+        notificationHandle: self.notificationId
+      }
+    });
+    self.saved = true;
+    return {sturdyRef: sturdyRef};
+  });
+};
+
+SandstormCoreImpl.prototype.restore = function (sturdyRef) {
+  var self = this;
+  return inMeteor(function () {
+    var hashedSturdyRef = hashSturdyRef(sturdyRef);
+    var token = ApiTokens.findOne(hashedSturdyRef);
+    if (!token) {
+      throw new Error("No token found to restore");
+    }
+    if (token.frontendRef) {
+      if (token.frontendRef.notificationHandle) {
+        var notificationId = token.frontendRef.notificationHandle;
+        return {cap: makeNotificationHandle(notificationId, true)};
+      } else {
+        throw new Error("Unknown frontend token type.");
+      }
+    } else if (token.objectId) {
+      return useGrain(self.grainId, function (supervisor) {
+        return supervisor.restore(token.objectId);
+      });
+    } else {
+      throw new Error("Unknown token type.");
+    }
+  });
+};
+
+SandstormCoreImpl.prototype.drop = function (sturdyRef) {
+  var grainId = this.grainId;
+  return inMeteor(function () {
+    var hashedSturdyRef = hashSturdyRef(sturdyRef);
+    var token = ApiTokens.findOne({_id: hashedSturdyRef});
+    if (!token) {
+      return;
+    }
+    if (token.frontendRef) {
+      if (token.frontendRef.notificationHandle) {
+        var notificationId = token.frontendRef.notificationHandle;
+        ApiTokens.remove({_id: hashedSturdyRef});
+        var anyToken = ApiTokens.findOne({"frontendRef.notificationHandle": notificationId});
+        if (!anyToken) {
+          // No other tokens referencing this notification exist, so dismiss the notification
+          dismissNotification(notificationId);
```

Comment:
In the case that the handle is dropped, we don't actually want to call cancel() on the underlying notification, but simply drop it as well. `cancel()` should be called when the user is actively asking to interrupt the task, not when the task merely declares itself done.